packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,346 @@
+
+		const intervalId = setInterval(async () => await respondToChat(), CHECK_FOR_RESPONSE_INTERVAL);
+
+		ws.once('close', () => {
+			ws.off('message', onMessage);
+			clearInterval(intervalId);
+			this.sessions.delete(key);
+		});
+
+		ws.on('message', onMessage);
+
+		const session: Session = {
+			connection: ws,
+			executionId,
+			sessionId,
+			intervalId,
+			isPublic: isPublic ?? false,
+			isProcessing: false,
+			lastHeartbeat: Date.now(),
+		};
+
+		this.sessions.set(key, session);
+
+		ws.send(N8N_HEARTBEAT);
+	}
+
+	private outgoingMessageHandler(sessionKey: string) {
+		return async () => {
+			let session: Session | undefined;
+			try {
+				session = this.sessions.get(sessionKey);
+
+				if (!session) return;
+				if (session.isProcessing) {
+					return;
+				}
+				session.isProcessing = true;
+
+				const { connection, executionId, sessionId, waitingNodeName, isPublic } = session;
+
+				if (!executionId || !connection) {
+					session.isProcessing = false;
+					return;
+				}
+
+				const execution = await this.getExecutionOrCleanupSession(executionId, sessionKey);
+
+				if (!execution) {
+					session.isProcessing = false;
+					return;
+				}
+
+				if (waitingNodeName) {
+					const lastNode = getLastNodeExecuted(execution);
+
+					if (execution.status === 'waiting' && lastNode?.name !== waitingNodeName) {
+						connection.send(N8N_CONTINUE);
+						session.waitingNodeName = undefined;
+					}
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'waiting') {
+					const message = getMessage(execution);
+
+					if (message !== undefined) {
+						connection.send(message);
+
+						const lastNode = getLastNodeExecuted(execution);
+
+						if (lastNode && shouldResumeImmediately(lastNode)) {
+							connection.send(N8N_CONTINUE);
+							const data: ChatMessage = { action: 'sendMessage', chatInput: '', sessionId };
+							await this.resumeExecution(executionId, data, sessionKey);
+							session.waitingNodeName = undefined;
+						} else {
+							session.waitingNodeName = lastNode?.name;
+						}
+					}
+
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'success') {
+					const shouldNotReturnLastNodeResponse =
+						!isPublic || (isPublic && isResponseNodeMode(execution));
+
+					if (shouldNotReturnLastNodeResponse) {
+						closeConnection(connection);
+						session.isProcessing = false;
+						return;
+					}
+
+					const textMessage = prepareMessageFromLastNode(execution);
+
+					connection.send(textMessage, () => {
+						closeConnection(connection);
+					});
+
+					session.isProcessing = false;
+					return;
+				}
+
+				session.isProcessing = false;
+			} catch (e) {
+				const error = ensureError(e);
+				if (session) session.isProcessing = false;
+				this.logger.error(
+					`Error sending message to chat in session ${sessionKey}: ${error.message}`,
+				);
+			}
+		};
```

Comment:
Using a `finally` block would simplify this and make sure `isProcessing` is always set to `false` when we exit the block